updating _compute_mask_indices fn to work with torch compile

Kirandevraj opened the pull request:
fixes #22849 

The inplace operations are replaced with out-of-place ones to fix the torch compile computational graph breakage.
This method converts the numpy operations to torch operations in _compute_mask_indices function.
The _compute_mask_indices is used when using SpecAugmentation in  wav2vec2 training.

@sanchit-gandhi

Kirandevraj committed:
updating _compute_mask_indices fn to work with torch compile

sanchit-gandhi commented on src/transformers/models/wav2vec2/modeling_wav2vec2.py:
```diff
@@ -199,8 +200,8 @@ def compute_num_masked_span(input_length):
-    spec_aug_mask = np.zeros((batch_size, sequence_length), dtype=bool)
-    spec_aug_mask_idxs = []
+    spec_aug_mask = torch.zeros((batch_size, sequence_length), dtype=torch.bool)
+    spec_aug_mask_idxs = torch.tensor([], dtype=torch.int64)
```
We're in `torch.int64` here

Kirandevraj replied:
updated to ```torch.int32``` here.

sanchit-gandhi commented on src/transformers/models/wav2vec2/modeling_wav2vec2.py:
```diff
@@ -1511,8 +1514,10 @@ def _mask_hidden_states(
-            mask_time_indices = torch.tensor(mask_time_indices, device=hidden_states.device, dtype=torch.bool)
-            hidden_states[mask_time_indices] = self.masked_spec_embed.to(hidden_states.dtype)
+            mask_time_indices = mask_time_indices.to(device=hidden_states.device)
+            temp_hidden_states = hidden_states.clone()
```
Also not sure why we `clone` here

Kirandevraj replied:
while using this code```hidden_states[mask_time_indices] = self.masked_spec_embed.to(hidden_states.dtype)``` I am facing this error:
Output 0 of CompiledFunctionBackward is a view and is being modified inplace. This view is the output of a 
function that returns multiple views. Such functions do not allow the output views to be modified inplace. You should replace the inplace operation by an out-of-place one.

sanchit-gandhi replied:
Okay interesting, but it works when we do:
```python
hidden_size[mask_time_indices] = self.masked_spec_embed.to(hidden_states.dtype)
```
on line 1505 above?

I wonder if we can open an issue on the PyTorch repo and see if we can get some clarify here? Not sure exactly what the best protocol is, but would ideally like to avoid cloning the entire tensor if possible

Kirandevraj replied:
The following code works:
```
temp_hidden_states = hidden_states.clone()
temp_hidden_states[mask_time_indices] = self.masked_spec_embed.to(hidden_states.dtype)
hidden_states = temp_hidden_states
```
 The following code throws above mentioned error:
```
hidden_states[mask_time_indices] = self.masked_spec_embed.to(hidden_states.dtype)
```
The following also throws the same error:
```
temp_hidden_states = hidden_states
temp_hidden_states[mask_time_indices] = self.masked_spec_embed.to(hidden_states.dtype)
hidden_states = temp_hidden_states
```

Can it be because the gradient for hidden_states has the hidden_size in the backward propagation graph and masked_spec_embed also has hidden_size in its backward propagation graph. And this could be an inplace operation?

sanchit-gandhi replied:
Quite possibly - that sounds like reasonable rationale for this behaviour. cc'ing @fxmarty - have you come across such an error when compiling with Torch 2.0? Torch is complaining that we're doing an index update for a Torch tensor in-place, but not really sure if cloning the tensor is the best option here, or whether we just need to make it non-contiguous?

fxmarty replied:
The issue looks to be that `temp_hidden_states[mask_time_indices] = self.masked_spec_embed.to(hidden_states.dtype)` is an inplace operation, and that autograd does not like inplace ops. Maybe [`torch.where`](https://pytorch.org/docs/stable/generated/torch.where.html) would work?

Kirandevraj replied:
```torch.where``` command is working with torch compile. 
The following can be used inplace of ```clone```.
```
expanded_embed = self.masked_spec_embed.unsqueeze(0).unsqueeze(0)
hidden_states = torch.where(mask_time_indices.unsqueeze(-1), expanded_embed, hidden_states)
```

gau-nernst replied:
I just saw this as @sanchit-gandhi mentioned this to me. I think using `Tensor.masked_fill()` can be slightly faster, while still being an out-of-place op?

Kirandevraj replied:
The ```Tensor.masked_fill()``` seems to only take ```float``` or ```Tensor``` value, I couldnot place ```embeded_mask```  ```1D``` vector in ```masked_fill```.

sanchit-gandhi reviewed (commented):
Looks really great already @Kirandevraj! Left some questions below (by and large for my own understanding).

We can just verify that the spec aug tests still pass:
https://github.com/huggingface/transformers/blob/bd9dfc23b9954b3b623425b31be181698d916a09/tests/models/wav2vec2/test_modeling_wav2vec2.py#L1240

And also add a new test to check that compiling the forward call works when we have spec aug activated

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_24668). All of your documentation changes will be reflected on that endpoint.

Kirandevraj committed:
Update src/transformers/models/wav2vec2/modeling_wav2vec2.py

Co-authored-by: Sanchit Gandhi <93869735+sanchit-gandhi@users.noreply.github.com>

Kirandevraj committed:
update src/transformers/models/wav2vec2/modeling_wav2vec2.py

Kirandevraj commented on src/transformers/models/wav2vec2/modeling_wav2vec2.py:
```diff
@@ -252,7 +251,8 @@ def compute_num_masked_span(input_length):
     # scatter indices to mask
-    np.put_along_axis(spec_aug_mask, spec_aug_mask_idxs, 1, -1)
+    mask_values = torch.ones(spec_aug_mask_idxs.shape, dtype=torch.bool)
+    spec_aug_mask = spec_aug_mask.scatter(1, spec_aug_mask_idxs, mask_values)
```
```scatter_``` is indeed easier.
We still need mask_values with True entries in the added code:
I will add this change
```
mask_values = torch.ones(spec_aug_mask_idxs.shape, dtype=torch.bool)
spec_aug_mask.scatter_(1, spec_aug_mask_idxs, mask_values)
```

Kirandevraj committed:
updating int32 and scatter_

Kirandevraj committed:
reverting clone function when mask_time_indices is not none

Kirandevraj commented on src/transformers/models/wav2vec2/modeling_wav2vec2.py:
```diff
@@ -1502,7 +1502,10 @@ def _mask_hidden_states(
         if mask_time_indices is not None:
             # apply SpecAugment along time axis with given mask_time_indices
-            hidden_states[mask_time_indices] = self.masked_spec_embed.to(hidden_states.dtype)
+            temp_hidden_states = hidden_states.clone()
```
This change may not be required as the mask_time_indices values will be given to the model. I have reverted it back to the previous version in 150a666bfca6416683c8acb3641dd06591ccd133

Kirandevraj commented on src/transformers/models/wav2vec2/modeling_wav2vec2.py:
```diff
@@ -227,22 +228,20 @@ def compute_num_masked_span(input_length):
-        spec_aug_mask_idx = np.concatenate(
-            [spec_aug_mask_idx, np.ones(max_num_masked_span - num_masked_span, dtype=np.int32) * dummy_mask_idx]
+        spec_aug_mask_idx = torch.cat(
+            (spec_aug_mask_idx, torch.ones(max_num_masked_span - num_masked_span, dtype=torch.int32) * dummy_mask_idx)
```
updated to int32

Kirandevraj committed:
updating _compute_mask_indices test cases by removing torch.from_numpy

Kirandevraj committed:
corrected incorrect variable name

Kirandevraj commented:
> And also add a new test to check that compiling the forward call works when we have spec aug activated

We already have a test case ```test_mask_time_prob_ctc``` that check the forward call with spec aug activated. 
Do you mean when using compile mode - we want to have a test case?

sanchit-gandhi commented:
> Do you mean when using compile mode - we want to have a test case?

Yes please - one test to make sure this PR gives the expected behaviour would be grand!

sanchit-gandhi commented:
I believe we just need an end-to-end test here and then we're good to go right @Kirandevraj?

Kirandevraj committed:
torch.where inplace of clone

Kirandevraj commented:
Yes, 
Also other test cases that expects ```np.ndarray``` output from the ```_compute_mask_indices``` have to updated to expect the new output ```torch.Tensor``` from ```_compute_mask_indices```

sanchit-gandhi commented:
Awesome - this is super close to completion then! Would you like to see it home? Feel free to ping me with any other questions / queries, more than happy to help!

Kirandevraj commented:
Yes, I am working on the test case. let me share it soon.

Kirandevraj commented:
When I use ```sampled_negative_indices``` in test cases. I am facing the following error from the last line.
```AssertionError('Loop-carried variable _tmp1 has initial type <[1, 2048], int1> but is re-assigned to  <[1, 2048], int8> in loop! Please make sure that the type stays consistent.')``` error by the 'inductor'. 
Sample code: 
```
import torch
from transformers import AutoFeatureExtractor, Wav2Vec2ForPreTraining
from transformers.models.wav2vec2.modeling_wav2vec2 import _compute_mask_indices, _sample_negative_indices
from datasets import load_dataset

if torch.cuda.is_available():
    device = torch.device("cuda")
    print("CUDA is available")
else:
    device = torch.device("cpu")
    print("CUDA is not available")


feature_extractor = AutoFeatureExtractor.from_pretrained("facebook/wav2vec2-base")
model = Wav2Vec2ForPreTraining.from_pretrained("facebook/wav2vec2-base")
model.to(device)

ds = load_dataset("hf-internal-testing/librispeech_asr_dummy", "clean", split="validation")
input_values = feature_extractor(ds[0]["audio"]["array"], return_tensors="pt").input_values.to(device)  # Batch size 1

# compute masked indices
batch_size, raw_sequence_length = input_values.shape
sequence_length = model._get_feat_extract_output_lengths(raw_sequence_length).item()
mask_time_indices = _compute_mask_indices(shape=(batch_size, sequence_length), mask_prob=0.2, mask_length=2)

sampled_negative_indices = _sample_negative_indices(
features_shape=(batch_size, sequence_length),
num_negatives=model.config.num_negatives,
mask_time_indices=mask_time_indices,
)

mask_time_indices = torch.tensor(data=mask_time_indices, device=input_values.device, dtype=torch.long)
sampled_negative_indices = torch.tensor(data=sampled_negative_indices, device=input_values.device, dtype=torch.long)

compiled_model = torch.compile(model)
outputs = compiled_model(input_values,mask_time_indices=mask_time_indices, sampled_negative_indices=sampled_negative_indices)
```
I believe this is from ```if sampled_negative_indices is not None:``` sub-block code snippet in forward function from wav2vec2. 
Should we go about fixing it to make it work with torch compile or can we create test case without using the ```sampled_negative_indices```
Can you direct me where the assertion error is getting raised to inquire.

sanchit-gandhi commented:
Do you have a full stack trace @Kirandevraj? This should give us a clue as to where the dtype is changing. IMO it could be worth going through the `sampled_negative_indices` function with a debugger and just checking that the input/output dtypes are consistent. The error message suggests we're doing an un-intentional upcast/downcast somewhere in the function

Kirandevraj committed:
adding test case torch.compile()

Kirandevraj commented:
The above error was from ```neg_is_pos.any()```. I updated the code to ```neg_is_pos.sum() > 0``` to work with torch compile.
I have added the test case ```test_torch_compiled_model_for_pretraining``` taking inspiration from ```test_model_for_pretraining```. I have updated few inplace operation with ```torch.where``` function.

sanchit-gandhi commented:
Awesome, sounds great! Let me know when you'd like me to take a second look at the code + tests! Getting close to merge here!

Kirandevraj commented:
These are the four test errors that I am facing in my brach and the main branch in my system and all the other test cases are passing from the test_modeling_wav2vec2.py
```
FAILED tests/models/wav2vec2/test_modeling_wav2vec2.py::Wav2Vec2ModelTest::test_initialization - AssertionError: 0.5709530115127563 not found in [0.0, 1.0] : Parameter wav2vec2.encoder.pos_conv_embed.conv.parametrizations.weight.original0 of model <class 'transformers.models.wav2vec2.modeling_wav2vec2.Wa...
FAILED tests/models/wav2vec2/test_modeling_wav2vec2.py::Wav2Vec2ModelTest::test_torch_fx - AssertionError: Couldn't trace module: symbolically traced variables cannot be used as inputs to control flow
FAILED tests/models/wav2vec2/test_modeling_wav2vec2.py::Wav2Vec2ModelTest::test_torch_fx_output_loss - AssertionError: Couldn't trace module: symbolically traced variables cannot be used as inputs to control flow
FAILED tests/models/wav2vec2/test_modeling_wav2vec2.py::Wav2Vec2RobustModelTest::test_initialization - AssertionError: 0.585796058177948 not found in [0.0, 1.0] : Parameter wav2vec2.encoder.pos_conv_embed.conv.parametrizations.weight.original0 of model <class 'transformers.models.wav2vec2.modeling_wav2vec2.Wav...
```
Can you take a look.

sanchit-gandhi commented:
Hey @Kirandevraj! The first two errors look like model initialisation errors (the dynamic range of the weights isn't in the expected range). These should be quite easy to fix as a last step, but I don't think we've changed anything in the state dict or the initialisation process, so shouldn't expect them to change?

It looks like the model is not working with torch fx tracing. Could you try running the model with trace enabled?

```python
from transformers import Wav2Vec2Model, Wav2Vec2FeatureExtractor
from transformers.utils.fx import symbolic_trace
import numpy as np

model = Wav2Vec2Model.from_pretrained("hf-internal-testing/tiny-random-wav2vec2")
feature_extractor = Wav2Vec2FeatureExtractor.from_pretrained("hf-internal-testing/tiny-random-wav2vec2")

raw_speech = np.ones(16000)

inputs = feature_extractor(raw_speech, sampling_rate=16000, return_tensors="pt")
input_names = list(inputs.keys())

traced_model = symbolic_trace(model, input_names)
traced_output = traced_model(inputs)
```
